### ISSUE TITLE
Code Review

### ISSUE THREAD

**Issue by jordons-hf** [Bethel-Eyo/HelloContacts#1]
Hi Bethel,
 
I’m Jordon, a sr. software developer at HelloFresh. I just finished reviewing your test! At HelloFresh, we encourage code reviews to give our applicants constructive feedback and discuss it interactively.
 
First of all, I want to thank you for the efforts you put into the test. To summarise my overview:
 
Kudos for:
 

- Great job with the project description (README). It explains the process of your implementation really well.
- Great job with extensive test coverage, especially with respect to how data affects the UI that is being rendered.
- Good to see the use of styled-components, we also use CSS-in-JS at HelloFresh.
- Excellent responsiveness and remarkably nice design, I really enjoyed reviewing this part and believe that attention to detail like this is very important for a frontend engineer

 
 
Points you could improve:
 

- The workflow for the solution made it difficult to read. 
  - There was just 1 commit which made it hard to follow the logical flow of your solution. 
  - Commits can be organized to better reflect the logical flow of your solution. I would recommend looking at the conventional commits project.
  - On the abstraction topic, I found that there was repeated usage and nesting of dispatched actions with additional side-effects from one layer to another. 
- You’ve included rules for linting but have not followed them.
  - Running the lint command produces 360 problems, 122 of which were potentially fixable by running the --fix command.
- The handling of contacts was not very performant.
  - In your actions you fetch all contacts for each operation.
  - This results in calling across the JS bridge into native APIs unnecessarily and is a performance issue.

 
I hope you have enjoyed working on this test and taking the review. This is a little demonstration of how we work at HelloFresh. I'm looking forward to hearing your thoughts on the points I mentioned in this review.
 
Please go through the couple of comments I left in this issue, and feel free to add any replies or ask any questions, however, please don't submit any further commits as we won't be able to review them.

**Comment by Bethel-Eyo** [comment]
Hi @jordons-hf,

Thanks for your feedback. I really appreciate it.

- In respect to the one 1 commit and push, I only created the repo and pushed as a medium of submission, but I will definitely take that into consideration next time.

- On the abstraction topic, it would be nice if you can provide me with ideas on how I would have gone about the abstraction or references I can look at. and just to break down the side effects a little further. This will be greatly appreciated.

- I will definitely pay more attention to the linting command and rules next time.

- For fetching the contacts on each operation, would it have been a better idea to just display the updated state every time there was a change and just sync when the user relaunches the app?

I would appreciate every constructive feedback I can get from this. Cheers

**Comment by jordons-hf** [comment]
Hey @Bethel-Eyo,

You're very welcome! Here are some additional thoughts based on your response:

Re: abstraction I would point to a couple of examples:

https://github.com/Bethel-Eyo/HelloContacts/blob/ce106b020935b970b66481738275aa0fb3c1d480/src/store/action-creators/index.ts#L103-L106

https://github.com/Bethel-Eyo/HelloContacts/blob/ce106b020935b970b66481738275aa0fb3c1d480/src/store/action-creators/index.ts#L115-L120

By coupling the result of these actions to additional side effects such as alerting and navigating we can observe a couple of things:

- We've introduced additional dependencies that could be removed to simplify and promote re-use (navigation, alert)
  - Here we could avoid coupling navigation and alert (and calling them directly) by accepting a callback that is executed upon successful update/create/delete/etc.
  - Alternatively we could consider async/await syntax to allow us to await the dispatch, then in-line from the calling screen/logic we could execute an alert or navigation call in a functional manner [See the async/await documentation here](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Asynchronous/Async_await) 
 - The cost of including these calls directly in our action creator is that we would need to refactor the code in order to call update/create/delete/etc. from another area in our app without encountering the alert/navigate side effect

Regarding contact fetching:

What you've proposed is certainly an option. I think the most important optimization to be made would be to not have to call getContacts for every operation. Using the contact's `id` field may be a source for maintaining your current workflow of creating/reading/updating/deleting them individually without sacrificing the support for synchronized contacts.

Thanks again for completing this project, it was a pleasure to discuss with you!
